File: iosp1/P1/P1/UI/Screens/DailyRoutineScreen.swift
import SwiftUI

private enum RoutineStepType {
    case photo, pain, vitals, done
}

struct DailyRoutineScreen: View {
    let followUpId: String
    let followUpTitle: String
    let rules: FollowUpRules?
    let onBack: () -> Void
    let onComplete: () -> Void

    @State private var steps: [RoutineStepType] = []
    @State private var stepIndex = 0

    // Collected measurements passed between steps
    @State private var painLevel: Double = 0
    @State private var tempValue: String = ""
    @State private var bpValue: String = ""
    @State private var hrValue: String = ""

    @State private var isSaving = false
    @State private var saveError: String? = nil

    var body: some View {
        VStack(spacing: 0) {
            LpmTopBar(title: followUpTitle, onBack: onBack)

            if !steps.isEmpty {
                VStack(spacing: 0) {
                    LpmStepIndicator(currentStep: stepIndex + 1, totalSteps: steps.count)
                        .padding(.top, 16)

                    Spacer().frame(height: 16)

                    if stepIndex < steps.count {
                        switch steps[stepIndex] {
                        case .photo:
                            PhotoStep(onPhotoTaken: { stepIndex += 1 })
                        case .pain:
                            PainStep(painLevel: $painLevel, onContinue: { stepIndex += 1 })
                        case .vitals:
                            VitalsStep(
                                rules: rules,
                                tempValue: $tempValue,
                                bpValue: $bpValue,
                                hrValue: $hrValue,
                                onContinue: { stepIndex += 1 }
                            )
                        case .done:
                            DoneStep(isSaving: isSaving, saveError: saveError, onFinish: onComplete)
                        }
                    }
                }
                .padding(.horizontal, 20)
            }
            Spacer()
        }
        .background(Color.white)
        .onAppear {
            var list: [RoutineStepType] = []
            if rules?.photos == true { list.append(.photo) }
            if rules?.pain != false { list.append(.pain) }
            if rules?.temperature == true || rules?.blood_pressure == true || rules?.smartwatch == true {
                list.append(.vitals)
            }
            list.append(.done)
            self.steps = list
        }
        .onChange(of: stepIndex) { newIndex in
            // When we reach the "done" step, post to API
            if newIndex < steps.count && steps[newIndex] == .done {
                saveToTimeline()
            }
        }
    }

    private func saveToTimeline() {
        isSaving = true
        saveError = nil

        Task {
            var lines: [String] = ["Routine Check-in:"]
            if steps.contains(.pain) {
                lines.append("• Pain Level: \(Int(painLevel))/10")
            }
            if !tempValue.isEmpty {
                lines.append("• Temperature: \(tempValue) °C")
            }
            if !bpValue.isEmpty {
                lines.append("• Blood Pressure: \(bpValue) mmHg")
            }
            if !hrValue.isEmpty {
                lines.append("• Heart Rate: \(hrValue) bpm")
            }
            if steps.contains(.photo) {
                lines.append("• Photo: taken")
            }

            let content = lines.joined(separator: "\n")
            let dateLabel = "Routine"

        Task {
            _ = TimelineRepository.addEvent(
                subscriptionId: followUpId,
                request: TimelineEventRequest(
                    content: content,
                    date_label: dateLabel,
                    effective_date: nil
                )
            )
            await MainActor.run { isSaving = false }
        }
        }
    }
}

// MARK: - Steps

private struct PhotoStep: View {
    let onPhotoTaken: () -> Void
    @State private var showCamera = false
    @State private var capturedImage: UIImage? = nil

    var body: some View {
        VStack(spacing: 20) {
            Text("Place the area to track inside the frame.")
                .font(.body)
                .foregroundColor(.gray)
                .multilineTextAlignment(.center)

            if let img = capturedImage {
                Image(uiImage: img)
                    .resizable()
                    .scaledToFill()
                    .frame(maxWidth: .infinity, maxHeight: 300)
                    .clipShape(RoundedRectangle(cornerRadius: 12))
            } else {
                ZStack {
                    RoundedRectangle(cornerRadius: 12)
                        .fill(Color(UIColor.systemGray6))
                        .frame(maxWidth: .infinity, maxHeight: 300)

                    VStack {
                        Image(systemName: "camera.viewfinder")
                            .font(.system(size: 48))
                            .foregroundColor(.gray)
                        Text("Tap to open camera")
                            .font(.callout)
                            .foregroundColor(.gray)
                            .padding(.top, 8)
                    }
                }
                .onTapGesture {
                    showCamera = true
                }
            }

            Spacer()

            LpmPrimaryButton(text: capturedImage == nil ? "Skip" : "Continue", action: onPhotoTaken)
        }
        .fullScreenCover(isPresented: $showCamera) {
            CameraView(selectedImage: $capturedImage)
                .edgesIgnoringSafeArea(.all)
        }
    }
}

private struct PainStep: View {
    @Binding var painLevel: Double
    let onContinue: () -> Void

    var body: some View {
        VStack(spacing: 24) {
            Text("Pain Assessment")
                .font(.system(size: 11, weight: .bold))
                .kerning(1.2)
                .foregroundColor(.gray)
                .frame(maxWidth: .infinity, alignment: .leading)

            Text("Pain level today")
                .font(.title3)
                .fontWeight(.semibold)
                .frame(maxWidth: .infinity, alignment: .leading)

            HStack {
                Text("0").foregroundColor(.gray).font(.caption)
                Spacer()
                Text("10").foregroundColor(.gray).font(.caption)
            }

            Slider(value: $painLevel, in: 0...10, step: 1)
                .tint(.black)

            Text("\(Int(painLevel)) / 10")
                .font(.title)
                .fontWeight(.bold)

            Spacer()
            LpmPrimaryButton(text: "Save", action: onContinue)
        }
    }
}

private struct VitalsStep: View {
    let rules: FollowUpRules?
    @Binding var tempValue: String
    @Binding var bpValue: String
    @Binding var hrValue: String
    let onContinue: () -> Void

    var body: some View {
        ScrollView {
            VStack(alignment: .leading, spacing: 24) {
                Text("Vitals Check-in")
                    .font(.system(size: 11, weight: .bold))
                    .kerning(1.2)
                    .foregroundColor(.gray)

                if rules?.temperature == true {
                    VStack(alignment: .leading, spacing: 12) {
                        Text("Quick Select Temperature")
                            .font(.caption)
                            .foregroundColor(.gray)

                        HStack {
                            ForEach(["36.5", "37.0", "37.5", "38.0"], id: \.self) { temp in
                                Button(action: { tempValue = temp }) {
                                    Text("\(temp)°")
                                        .font(.subheadline)
                                        .padding(.vertical, 8)
                                        .padding(.horizontal, 16)
                                        .background(tempValue == temp ? Color.black : Color(UIColor.systemGray6))
                                        .foregroundColor(tempValue == temp ? .white : .black)
                                        .cornerRadius(20)
                                }
                            }
                        }

                        TextField("Or enter specific (°C)", text: $tempValue)
                            .keyboardType(.decimalPad)
                            .padding()
                            .background(Color(UIColor.systemGray6))
                            .cornerRadius(8)
                    }
                }

                if rules?.blood_pressure == true {
                    TextField("Blood Pressure (mmHg)", text: $bpValue)
                        .padding()
                        .background(Color(UIColor.systemGray6))
                        .cornerRadius(8)
                }

                if rules?.smartwatch == true {
                    TextField("Heart Rate (bpm)", text: $hrValue)
                        .keyboardType(.numberPad)
                        .padding()
                        .background(Color(UIColor.systemGray6))
                        .cornerRadius(8)
                }

                Spacer().frame(height: 40)
                LpmPrimaryButton(text: "Save", action: onContinue)
            }
        }
    }
}

private struct DoneStep: View {
    let isSaving: Bool
    let saveError: String?
    let onFinish: () -> Void

    var body: some View {
        VStack(spacing: 24) {
            Spacer()

            if isSaving {
                ProgressView()
                    .scaleEffect(1.5)
                    .padding(.bottom, 8)
                Text("Saving your data...")
                    .font(.body)
                    .foregroundColor(.gray)
            } else {
                Circle()
                    .fill(Color.black)
                    .frame(width: 80, height: 80)
                    .overlay(
                        Image(systemName: "checkmark")
                            .font(.system(size: 36, weight: .bold))
                            .foregroundColor(.white)
                    )

                Text("Routine saved")
                    .font(.system(size: 11, weight: .bold))
                    .kerning(1.2)
                    .foregroundColor(.gray)
                    .padding(.top, 16)

                if let err = saveError {
                    Text(err)
                        .font(.caption)
                        .foregroundColor(.orange)
                        .multilineTextAlignment(.center)
                }

                Text("Today's data has been saved. Come back tomorrow for your next routine.")
                    .font(.body)
                    .foregroundColor(.black)
                    .multilineTextAlignment(.center)
            }

            Spacer()

            LpmPrimaryButton(text: "Back to home", action: onFinish)
                .disabled(isSaving)
                .opacity(isSaving ? 0.5 : 1)
            Spacer()
        }
    }
}

File: androidp1/app/src/main/java/com/preappointment1/app/ui/screens/DailyRoutineScreen.kt
package com.preappointment1.app.ui.screens

import androidx.compose.foundation.background
import androidx.compose.foundation.layout.*
import androidx.compose.foundation.verticalScroll
import androidx.compose.foundation.rememberScrollState
import androidx.compose.foundation.shape.CircleShape
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.filled.Check
import androidx.compose.material3.*
import androidx.compose.runtime.*
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.text.style.TextAlign
import androidx.compose.foundation.text.KeyboardOptions
import androidx.compose.ui.draw.clip
import androidx.compose.ui.text.input.KeyboardType
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import com.preappointment1.app.data.model.FollowUpRules
import com.preappointment1.app.ui.components.*
import com.preappointment1.app.ui.theme.*

private enum class RoutineStepType {
    Photo, Pain, Vitals, Done
}

@Composable
fun DailyRoutineScreen(
    followUpTitle: String = "Daily Routine",
    rules: FollowUpRules?,
    onBack: () -> Unit,
    onComplete: () -> Unit,
    modifier: Modifier = Modifier
) {
    val steps = remember(rules) {
        val list = mutableListOf<RoutineStepType>()
        if (rules?.photos == true) list.add(RoutineStepType.Photo)
        if (rules?.pain != false) list.add(RoutineStepType.Pain)
        if (rules?.temperature == true || rules?.bloodPressure == true || rules?.smartwatch == true) {
            list.add(RoutineStepType.Vitals)
        }
        list.add(RoutineStepType.Done)
        list
    }

    var stepIndex by remember { mutableIntStateOf(0) }
    var photoFilename by remember { mutableStateOf<String?>(null) }

    Scaffold(
        topBar = { LpmTopBar(title = followUpTitle, onBack = onBack) },
        containerColor = CanvasBackground,
        modifier = modifier
    ) { padding ->
        Column(
            modifier = Modifier
                .fillMaxSize()
                .padding(padding)
                .padding(horizontal = 22.dp)
        ) {
            Spacer(modifier = Modifier.height(16.dp))
            LpmStepIndicator(currentStep = stepIndex + 1, totalSteps = steps.size)
            Spacer(modifier = Modifier.height(20.dp))

            when (steps[stepIndex]) {
                RoutineStepType.Photo -> PhotoStep(onPhotoTaken = { filename ->
                    photoFilename = filename
                    stepIndex++
                })
                RoutineStepType.Pain -> PainStep(onContinue = { stepIndex++ })
                RoutineStepType.Vitals -> VitalsStep(rules = rules, onContinue = { stepIndex++ })
                RoutineStepType.Done -> DoneStep(onFinish = onComplete)
            }
        }
    }
}

@Composable
private fun PhotoStep(onPhotoTaken: (String?) -> Unit) {
    Column(modifier = Modifier.fillMaxSize()) {
        var photoSaved by remember { mutableStateOf<String?>(null) }

        if (photoSaved == null) {
            MeasurementPhotoCapture(
                onPhotoCaptured = { fileName ->
                    photoSaved = fileName
                },
                previewHeight = 280
            )
        } else {
            Box(
                modifier = Modifier
                    .fillMaxWidth()
                    .height(120.dp)
                    .clip(RoundedCornerShape(20.dp))
                    .background(MintBadge),
                contentAlignment = Alignment.Center
            ) {
                Text("✓ Photo saved", color = MintBadgeText, fontWeight = FontWeight.Bold, fontSize = 16.sp)
            }
            Spacer(modifier = Modifier.height(20.dp))
            LpmPrimaryButton(text = "Continue", onClick = { onPhotoTaken(photoSaved) })
        }
    }
}

@Composable
private fun PainStep(onContinue: () -> Unit) {
    var painLevel by remember { mutableFloatStateOf(0f) }

    Column(modifier = Modifier.fillMaxSize()) {
        LpmSectionTitle("Pain Assessment")
        Spacer(modifier = Modifier.height(24.dp))

        Text(
            text = "Pain level today",
            style = MaterialTheme.typography.titleMedium,
            fontWeight = FontWeight.Bold,
            color = TextPrimary
        )
        Spacer(modifier = Modifier.height(8.dp))
        Row(
            modifier = Modifier.fillMaxWidth(),
            horizontalArrangement = Arrangement.SpaceBetween
        ) {
            Text("0 (None)", color = TextSecondary, style = MaterialTheme.typography.bodySmall)
            Text("10 (Severe)", color = TextSecondary, style = MaterialTheme.typography.bodySmall)
        }
        Slider(
            value = painLevel,
            onValueChange = { painLevel = it },
            valueRange = 0f..10f,
            steps = 9,
            colors = SliderDefaults.colors(
                thumbColor = SagePrimary,
                activeTrackColor = SagePrimary,
                inactiveTrackColor = MintBadge
            )
        )
        Text(
            text = "${painLevel.toInt()} / 10",
            modifier = Modifier.fillMaxWidth(),
            textAlign = TextAlign.Center,
            style = MaterialTheme.typography.titleLarge,
            fontWeight = FontWeight.ExtraBold,
            color = SagePrimary
        )

        Spacer(modifier = Modifier.weight(1f))
        LpmPrimaryButton(text = "Save & Continue", onClick = onContinue)
        Spacer(modifier = Modifier.height(24.dp))
    }
}

@Composable
private fun VitalsStep(rules: FollowUpRules?, onContinue: () -> Unit) {
    var tempValue by remember { mutableStateOf("") }
    var bpValue by remember { mutableStateOf("") }
    var hrValue by remember { mutableStateOf("") }

    Column(modifier = Modifier.fillMaxSize()) {
        Column(
            modifier = Modifier
                .weight(1f)
                .verticalScroll(rememberScrollState())
        ) {
            LpmSectionTitle("Vitals Check-in")
            Spacer(modifier = Modifier.height(24.dp))

            if (rules?.temperature == true) {
                Text("Quick Select Temperature", style = MaterialTheme.typography.bodySmall, color = TextSecondary)
                Spacer(modifier = Modifier.height(8.dp))
                Row(
                    modifier = Modifier.fillMaxWidth(),
                    horizontalArrangement = Arrangement.SpaceBetween
                ) {
                    listOf("36.5", "37.0", "37.5", "38.0").forEach { temp ->
                        FilterChip(
                            selected = tempValue == temp,
                            onClick = { tempValue = temp },
                            label = { Text("$temp°") },
                            colors = FilterChipDefaults.filterChipColors(
                                selectedContainerColor = SagePrimary,
                                selectedLabelColor = Color.White
                            )
                        )
                    }
                }
                Spacer(modifier = Modifier.height(8.dp))
                OutlinedTextField(
                    value = tempValue,
                    onValueChange = { tempValue = it },
                    label = { Text("Or enter specific (°C)") },
                    modifier = Modifier.fillMaxWidth().padding(bottom = 16.dp),
                    shape = RoundedCornerShape(16.dp),
                    singleLine = true,
                    keyboardOptions = KeyboardOptions(keyboardType = KeyboardType.Number)
                )
            }

            if (rules?.bloodPressure == true) {
                OutlinedTextField(
                    value = bpValue,
                    onValueChange = { bpValue = it },
                    label = { Text("Blood Pressure (mmHg)") },
                    modifier = Modifier.fillMaxWidth().padding(bottom = 16.dp),
                    shape = RoundedCornerShape(16.dp),
                    singleLine = true,
                    keyboardOptions = KeyboardOptions(keyboardType = KeyboardType.Number)
                )
            }

            if (rules?.smartwatch == true) {
                OutlinedTextField(
                    value = hrValue,
                    onValueChange = { hrValue = it },
                    label = { Text("Heart Rate (bpm)") },
                    modifier = Modifier.fillMaxWidth().padding(bottom = 16.dp),
                    shape = RoundedCornerShape(16.dp),
                    singleLine = true,
                    keyboardOptions = KeyboardOptions(keyboardType = KeyboardType.Number)
                )
            }
        }
        
        LpmPrimaryButton(text = "Save & Continue", onClick = onContinue)
        Spacer(modifier = Modifier.height(24.dp))
    }
}

@Composable
private fun DoneStep(onFinish: () -> Unit) {
    Column(
        modifier = Modifier.fillMaxSize(),
        horizontalAlignment = Alignment.CenterHorizontally,
        verticalArrangement = Arrangement.Center
    ) {
        Box(
            modifier = Modifier
                .size(72.dp)
                .clip(CircleShape)
                .background(MintBadge),
            contentAlignment = Alignment.Center
        ) {
            Icon(
                Icons.Default.Check,
                contentDescription = null,
                tint = SagePrimary,
                modifier = Modifier.size(36.dp)
            )
        }
        Spacer(modifier = Modifier.height(24.dp))
        LpmSectionTitle("Routine Completed")
        Spacer(modifier = Modifier.height(12.dp))
        LpmBodyText(
            "Today's data has been securely saved to your timeline file.",
            modifier = Modifier.padding(horizontal = 20.dp)
        )
        Spacer(modifier = Modifier.height(36.dp))
        LpmPrimaryButton(
            text = "Back to Home",
            onClick = onFinish,
            modifier = Modifier.padding(horizontal = 24.dp)
        )
    }
}
